---
title: "EpiXprS: A poly-CpG site approach to gene expression prediction and association"
author: "Ian Loveless"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{EpiXprS: A poly-CpG site approach to gene expression prediction and association}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


DNA methylation has been shown to be in important regulator of gene expression. 
This relationship is complex, with multiple CpG sites in a  region influencing expression rather than by a single CpG site in isolation. The `EpiXprS` package provides a means for constructing these multi-CpG models, allowing for the further complexity of context dependent modifiers of the relationship between methylation and gene expression. In particular, `EpiXprS` was developed to quantify and account for the context dependent effect of race through the inclusion of CpG-by-race interaction terms. Namely, to construct the original models, paired DNA methylation and gene expression were downloaded from The Cancer Genome Atlas TCGA, using the `TCGAbiolinks` package, to assess the role of race as a modifier in the multivariate relationship between DNA methylation and gene expression. Eight tumor types were downloaded that had high African American representation to allow us to assess these potential differences. While the subsequent description therefore focuses on race (African American and European American) the context being evaluated, the contribution of any potential dichotomous modifier (e.g. environmental exposure) could be assessed. 


# Installation

To install this package, start `R` and enter:

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("EpiXprS")
```

This will install the package and all missing dependencies.



```{r setup}
library(EpiXprS)
set.seed(1014)

brca <- curatedTCGAData::curatedTCGAData("BRCA", c("RNASeqGene", "Methylation_methyl450"), dry.run=FALSE, version = '2.0.1') 
brcamatched <- intersectColumns(brca)
brcamatched <- brcamatched[,brcamatched$race %in% c('white','black or african american')]
coldat <- as(colData(brcamatched),'data.frame')
Methy <- assays(brcamatched)$`BRCA_Methylation_methyl450-20160128`[,substr(colnames(assays(brcamatched)$`BRCA_Methylation_methyl450-20160128`),14,15) == '01']
Counts <- assays(brcamatched)$`BRCA_RNASeqGene-20160128`[c(1:2),substr(colnames(assays(brcamatched)$`BRCA_RNASeqGene-20160128`),14,15) == '01']
Methy <- Methy[,which(substr(colnames(Methy),1,16) %in% substr(colnames(Counts),1,16))]
Counts <- Counts[,match(substr(colnames(Methy),1,16),substr(colnames(Counts),1,16))]
coldat <- coldat[match(substr(colnames(Methy),1,12),coldat$patientID),]
```

The main functionality of this package is to allow one to associate many CpG sites with gene expression, while accounting for possible race-specific effects. This is managed by including multiplicative interaction terms between race and CpG sites in a penalized regression model. The function `Construct()` 
has three methods, `Specific` which constructs models that ignore possible race-specific effects, `Adjusted` where race is included in the model, and `Interaction` where the aforementioned multiplicative interaction terms are included as well. The function returns 1) the prediction model, 2) the cross-validated predicted expression, 3) the percent of variability explained, 4) & 5) the race specific variability
explained for the `Adjusted` and `Interaction` model types, and 6) ensembl annotation for the genes modeled. 

```{r Construct Demonstration}
out <- Construct(x = Methy[,1:100], y = Counts[,1:100], clinical = coldat$race[1:100],
                           method ='Interaction', dist = NULL, nfolds = NULL, 
                          impute = TRUE, beta = TRUE, parallel = FALSE,
                 Methylation.Annotation = annotation_450k, Gene.Annotation = annotated_RNA)

rbind(head(out[1,1][[1]][[1]]),tail(out[1,1][[1]][[1]]))
head(out[1,2][[1]][[1]])
out[1,3][[1]]
out[1,4][[1]]
out[1,5][[1]]
out[1,6][[1]]
```

Another useful function, `ModelSearch()`, allows one to search through pre-constructed models to identify gene models in which a given CpG site is included. It also provides functionality to search for specific genes to identify the CpG sites that are included its model. 


```{r model_search, eval = FALSE}
library(ExperimentHub)
eh = ExperimentHub()
x <- query(eh, 'EpiXprS')
Cancer <- x[["EH6022"]]
ModelSearch('cg21837192', Cancer = Cancer)
```

Similarly, pre-constructed models can be used to predict gene expression from a  450K DNA methylation data set, using the `Predict()` function. Prior to use, we advise missing DNA methylation be imputed using the `Imputation()` function. This function is borrowed from the `methyLImp` package, to provide complete predicted expression. For more details, see `?Predict`. 

Finally, predicted expression scores from the `Predict()` function can be used to associate predicted expression with a phenotype of interest using the `Assoc` function. This function currently implements linear regression, logistic regression, and Cox proportional hazards models. For more details, see `?Assoc`.

```{r Session Information}
sessionInfo()
```
